sys/net/application_layer/unicoap/include/private/messaging.h

```diff
@@ -0,0 +1,341 @@
+ */
+int unicoap_messaging_process_rfc7252(const uint8_t* pdu, size_t size, bool truncated,
+                                      unicoap_packet_t* packet);
+
+/**
+ * @brief Forwards packet to messaging layer and the corresponding driver for further message
+ * processing, encoding, and transport I/O.
+ *
+ * The driver needs to encode the given message into a serialized PDU and use its own means
+ * of transmitting the packet to the packet's remote endpoint.
+ *
+ * This function calls the individual driver messaging implementation.
+ *
+ * @param[in,out] packet Packet to send
+ * @param flags Messaging flags
+ */
+int unicoap_messaging_send(unicoap_packet_t* packet, unicoap_messaging_flags_t flags);
+
+/** @brief Sends CoAP over UDP or DTLS packet, see @ref unicoap_messaging_send */
+int unicoap_messaging_send_rfc7252(unicoap_packet_t* packet, unicoap_messaging_flags_t flags);
+/**
+ * @brief Releases all buffers and state allocated in connection with the given endpoint
+ *
+ * Call this function from your driver if you encounter a severe error with a given endpoint
+ *
+ * @warning Do not call this function from within the exchange layer as messaging-layer state
+ * will not be released by this function.
+ *
+ * @param[in] endpoint Remote endpoint to release buffers for
+ */
+void unicoap_exchange_forget_endpoint(const unicoap_endpoint_t* endpoint);
+
+/**
+ * @brief Generates new token
+ *
+ * Token will be @ref CONFIG_UNICOAP_GENERATED_TOKEN_LENGTH bytes long.
+ *
+ * @param[in,out] token Buffer for generated token, must have a
+ * minimum capacity of @ref CONFIG_UNICOAP_GENERATED_TOKEN_LENGTH
+ */
+void unicoap_generate_token(uint8_t* token);
+
+/**
+ * @brief Retrieves the part of a resource flags bitfield relevant for the messaging driver.
+ *
+ * @param resource_flags Resource flags
+ * @return Messaging flags extracted from the given bitfield
+ */
+static inline unicoap_messaging_flags_t
+_messaging_flags_resource(unicoap_resource_flags_t resource_flags)
+{
+    /* We documented other flags are RFU, hence downcasting to the messaging
+     flags bitfield width is fine here */
+    return (unicoap_messaging_flags_t)resource_flags;
+}
+/** @}  */
+
+/* MARK: - Private Exchange-Layer Server API */
+/**
+ * @name Server
+ * @{
+ */
+/**
+ * @brief Informs the unicoap server a new packet has been preprocessed and can be handled.
+ *
+ * This function handles block-wise transfers.
+ *
+ * @param[in,out] packet Packet that will be processed by the server
+ * @param[in,out] resource Resource
+ *
+ * @return `0` on success
+ * @returns Negative errno on failure
+ */
+int unicoap_server_process_request(unicoap_packet_t* packet, const unicoap_resource_t* resource);
+
+/**
+ * @brief Sends entire response body, may be split into parts and then sent
+ *
+ * @param packet Response packet to send
+ * @param resource mandatory resource which is sending this response
```

Comment:
```suggestion
 * @param[in] packet Response packet to send
 * @param[in] resource mandatory resource which is sending this response
```

Is `packet` only `in`?

Reply:
No, the messaging implementation residing in the driver could mutate the message before sending it. this is intended behavior. As C lacks memory ownership, there is no way to express consuming access patterns with official language semantics.